File: src/main/java/leetcode/lists/offerII/Offer49.java
package leetcode.lists.offerII;

import common.*;

/**
 * @author wujingxinit@outlook.com
 * @date 2023/7/25 16:29
 */
public class Offer49 {

    int sum = 0;

    public int sumNumbers(TreeNode root) {
        if (root == null) {
            return 0;
        }
        dfs(root, 0);
        return sum;

    }

    private void dfs(TreeNode root, int val) {
        if (root == null) {
            return;
        }
        int s = (val * 10 + root.val);
        if (root.left == null && root.right == null) {
            sum += s;
            return;
        }
        dfs(root.left, s);
        dfs(root.right, s);
    }
}

File: go/src/leetcode/offerII/Offer49.go
package offerII

/**
 * @author wujingxinit@outlook.com
 * @date 2023/1/16 17:11
 */

func sumNumbers(root *TreeNode) int {
	var dfs func(root *TreeNode, prevSum int) int
	dfs = func(root *TreeNode, prevSum int) int {
		if root == nil {
			return 0
		}
		sum := prevSum*10 + root.Val
		if root.Left == nil && root.Right == nil {
			return sum
		}
		return dfs(root.Left, sum) + dfs(root.Right, sum)
	}
	return dfs(root, 0)
}
